torch/_inductor/mkldnn.py

```diff
@@ -789,47 +685,29 @@ def fuse_binary(gm: torch.fx.GraphModule):
+def convert_outplace_to_inplace(gm: torch.fx.GraphModule):
+    if not (torch.backends.mkldnn.enabled and torch.backends.mkldnn.is_available()):
+        return gm
+    # This function is about replace outplace with inplace for better performance(external call),
+    # which happen after AOTAutograd.
     for node in gm.graph.nodes:
-        if check_node_is_add_inplace(node) and check_binary_op_kwargs_is_default(node):
-            for (
-                node_kind,
-                fuse_func,
-            ) in computation_op_binary_op_fusion_inplace_map.items():
-                if not isinstance(node.args[0], torch.fx.Node) or not isinstance(
-                    node.args[1], torch.fx.Node
-                ):
-                    continue
-                if not binary_inputs_meta_is_same(node):
-                    continue
-                if check_node_kind(node.args[1], modules, node_kind):
-                    if len(node.args[1].users) > 1:
-                        continue
-                    # make sure the output and input are not same tensor.
-                    if node.args[1].args[0] == node.args[0]:
-                        continue
-                    computation_node = modules[node.args[1].target]
-                    if computation_node.training:
-                        continue
-                    # TODO: support padding str input("valid", "same").
-                    if type(computation_node) in [nn.Conv2d] and isinstance(
-                        computation_node.padding, str
-                    ):
-                        continue
-                    replace_and_fuse_for_binary(
-                        computation_node,
-                        node,
-                        fuse_func,
-                        "add",
-                        modules,
-                        1,  # conv module index
-                        0,  # binary op index
-                    )
-                    # Make sure the fused node is post node of node's inputs nodes.
-                    node.append(node.args[1])
-                    gm.graph.erase_node(node)
-                    break
+        if node.op == "call_function" and node.target in [
+            torch.ops.mkldnn._convolution_pointwise.binary
+        ]:
+            # args[0] and args[1] is _convolution_pointwise.binary's input,
+            # need to check whether args[1] can be written or not.
+            if node.args[1].op in ["placeholder", "output"]:
+                continue
+            # TODO: node.args[1].users > 1, but node.args[1] never be used after current node.
+            if len(node.args[1].users) > 1:
+                continue
+            if node.args[1] == node.args[0]:
+                continue
+            binary_attr = node.args[8]
+            unary_attr = node.args[10]
+            if binary_attr != "add" or unary_attr not in ["", "relu"]:
+                continue
+            node.target = torch.ops.mkldnn._convolution_pointwise_.binary
```

Comment:
Does the IR node for this op properly mark the mutation?

I worry inductor will try to reuse input buffer (which is now alised by the return value).

Reply:
Yes, it is marked mutation on creation: https://github.com/pytorch/pytorch/blob/master/torch/_inductor/ir.py#L3659